File: ex4/src/client/src/GameFlow.cpp
/**
 * # Ori Cohen
# ID: 207375783
# Yana Patyuk
# ID:317106755
 */

#include "../include/GameFlow.h"

GameFlow::GameFlow(LogicInterface *logic, Board *board, PlayerInterface *player1, PlayerInterface *player2, bool server) {
  this->logic = logic;
  this->board = board;
  this->currentTurn = EnumDeclration::X;
  this->player[0] = player1;
  this->player[1] = player2;
  this->clentServer = server;
  // some initialization
  board->SetCellStatus(board->NumOfRows() / 2, board->NumOfCols()/2, EnumDeclration::O);
  board->SetCellStatus(board->NumOfRows() / 2 + 1, board->NumOfCols() / 2 + 1, EnumDeclration::O);
  board->SetCellStatus(board->NumOfRows() / 2 + 1, board->NumOfCols() / 2, EnumDeclration::X);
  board->SetCellStatus(board->NumOfRows() / 2, board->NumOfCols() / 2 + 1, EnumDeclration::X);
}
void GameFlow::Run() {
  Board *b = this->board;
  int tunrnConter = 0;
  // run the game while it is still not over
	b->Print();
  while (!GameOver()) {
	  //check who's turn it is.
	/*  if(tunrnConter%2 == 1) {
		  this->currentTurn = EnumDeclration::O;
	  } else {
		  this->currentTurn = EnumDeclration::X;
	  }*/
      // if player has possible slots to place.
   if (this->logic->SlotsToPlace(this->player[tunrnConter % 2]->getEnumSymbol()).size() != 0) {
       // player makes a move.
	   if(clentServer && this->player[tunrnConter % 2] == NULL) continue;
     this->player[tunrnConter % 2]->MakeAMove(this->board, this->logic);
		if(!(clentServer && tunrnConter % 2 == 0)) b->Print();
   } else {
	   //in server-client mode: we have to send lastmove and recive nomove ene if
	   //we know other player cant move.
	   if(clentServer && (tunrnConter % 2 == 0)) {
		   this->player[tunrnConter % 2]->MakeAMove(this->board, this->logic);
		   tunrnConter++;
		   continue;
	   }
        // it doesn't have possible slots to place tag at
        // the turn passes over
    	  cout<< player[tunrnConter % 2]->GetSymbol()
              << " I'ts your move. but unfortunately you don't have anything to do," <<
             "therefore it's only fair that the play passes back to " <<
			 	 	 	 player[(tunrnConter + 1) % 2]->GetSymbol() << endl;
    	  this->board->SetLastMove("NoMove");
    	  //if(clentServer) this->player[tunrnConter % 2]->MakeAMove(NULL, NULL);
      }
   tunrnConter++;
   //this->board->Print();//print the board.
  }
  //print end game screen.
  EndGame();
}

bool GameFlow::GameOver() {
  if (this->logic->SlotsToPlace(EnumDeclration::X).size() != 0
      || this->logic->SlotsToPlace(EnumDeclration::O).size() != 0) {
    return false;
  }//if its the end of board and we made last move- send to server end.
  if(this->board->GetLastMove().compare("End") != 0) {
	  if(this->clentServer) this->player[0]->MakeAMove(this->board, this->logic);
	  this->board->SetLastMove("End");
	  if(this->clentServer) this->player[0]->MakeAMove(this->board, this->logic);
  }
  return true;
}

void GameFlow::EndGame() {
	  Board *b = this->board;

    int num_of_slots_in_board = b->NumOfRows() * b->NumOfCols();
    int num_of_used_slots = b->GetXSlots().size() + b->GetOSlots().size();
    if (num_of_used_slots >= num_of_slots_in_board) {
      cout << "Game is over, board is completely full" << endl;
    } else {
      cout << "Game is over, both sides can't make any more moves" << endl;
    }
    // declare result
    if (b->GetXSlots().size() > b->GetOSlots().size()) {
      cout << "X is the WINNER!";
    } else if (b->GetXSlots().size() < b->GetOSlots().size()) {
      cout << "O is the WINNER!";
    } else {
      cout << "It's a tie";
    }
}

File: ex6GUI/src/GameFlow.java

public class GameFlow {
	
	private  Logic logic;
	private	 Board board;
	private	 Player player[];
	private  CellStatus currentTurn;
	private int turnCounter;
	private GUI drawer;
			/**
			 * Class that holds all the information in order to run the game
			 * @param logic - logic the game run with
			 * @param board - game board
			 * @param player1
			 * @param player2
			 * @param drawer 
			 */
	public  GameFlow(Logic logic, Board board, Player player1, Player player2, GUI drawer) {
	    this.logic = logic;
	    this.board = board;
	    this.turnCounter = 0;
	    //this.currentTurn = CellStatus.X;
	    this.player = new Player[2];
	    this.player[0] = player1;
	    this.player[1] = player2;
	    this.drawer = drawer;
	    // some initialization
	    board.SetCellStatus(board.NumOfRows() / 2, board.NumOfCols() / 2, CellStatus.O);
	    board.SetCellStatus(board.NumOfRows() / 2 + 1, board.NumOfCols() / 2 + 1, CellStatus.O);
	    board.SetCellStatus(board.NumOfRows() / 2 + 1, board.NumOfCols() / 2, CellStatus.X);
	    board.SetCellStatus(board.NumOfRows() / 2, board.NumOfCols() / 2 + 1, CellStatus.X);
	}
		  /**
		   * Run the game
		   */
	public	  void Run() {
	/*	  Board b = this.board;
		  int tunrnConter = 0;
		  b.Print();
		  // run the game while it is still not over
		  while (!GameOver()) {
			  //check who's turn it is.
			  System.out.print("here3\n");
			  if(tunrnConter%2 == 1) {
				  this.currentTurn = CellStatus.O;
			  } else {
				  this.currentTurn = CellStatus.X;
			  }
		      // if player has possible slots to place.
		   if (this.logic.SlotsToPlace(this.currentTurn).size() != 0) {
		       // player makes a move.
		     this.player[tunrnConter % 2].MakeAMove(this.board, this.logic);
		   } else {
		        // it doesn't have possible slots to place tag at
		        // the turn passes over
			   System.out.print(player[tunrnConter % 2].GetSymbol() +
		              " I'ts your move. but unfortunately you don't have anything to do," +
		             "therefore it's only fair that the play passes back to other player"
		             );
		      }
		   tunrnConter++;
		   this.board.Print();//print the board.
		  }
		  System.out.print("here3\n");
		  //print end game screen.
		  EndGame();**/
	}
		  /**
		   * @return if Game is Over
		   */
	public	  boolean GameOver() {
		  if (this.logic.SlotsToPlace(CellStatus.X).size() != 0
			      || this.logic.SlotsToPlace(CellStatus.O).size() != 0) {
			    return false;
			  }
			  return true;
	}
		  /**
		   * return who is the winner by string.
		   */
	public  String EndGame() {
		  Board b = this.board;
		    // declare result
		    if (b.GetXSlots().size() > b.GetOSlots().size()) {
		    	return "Player 1";
		    } else if (b.GetXSlots().size() < b.GetOSlots().size()) {
		    	return "player 2";
		    } else {
		    	return "It's a TIE!";
		    }
	}
	/**
	 * play players turn. if slot illigal-return false and don't change the turn;
	 * @param slot
	 * @return
	 */
		public boolean playOneTurn(Slot slot) {
			 
			boolean correctMove = this.player[this.turnCounter % 2].MakeAMove(this.board, this.logic, slot);
			if(correctMove) this.turnCounter++;
			if(this.logic.SlotsToPlace(this.player[this.turnCounter % 2].GetEnumSymbol()).size() == 0) this.drawer.drawMessage("next!");
			this.drawer.drawPossibleMovesTitle(this.logic.SlotsToPlace(this.player[this.turnCounter % 2].GetEnumSymbol()));
			this.drawer.SetPlayer(this.player[(this.turnCounter ) % 2].GetEnumSymbol());
			this.drawer.drawGame();
			return correctMove;
		}
		/**
		 * @return board.
		 */
		public Board GetBoard() {
			return this.board;
		}
}
